File: examples/backend/python/app/projects.py
import json
from datetime import UTC, datetime, timedelta
from hashlib import sha256
from typing import cast
from uuid import UUID, uuid4

from fastapi import APIRouter, Depends, Header, Query, Response
from pydantic import BaseModel, ConfigDict, Field
from sqlalchemy import and_, or_, select, update
from sqlalchemy.exc import IntegrityError
from sqlalchemy.ext.asyncio import AsyncSession

from .auth import Principal, current_principal
from .database import database_session
from .models import IdempotencyKey, Project
from .problems import ApiProblem

router = APIRouter(prefix="/projects", tags=["projects"])


class ProjectInput(BaseModel):
    model_config = ConfigDict(extra="forbid")
    name: str = Field(min_length=2, max_length=120)
    description: str | None = Field(default=None, max_length=4000)


class ProjectUpdate(ProjectInput):
    version: int = Field(ge=1)


IDEMPOTENCY_SCOPE = "projects.create"


def normalized_project_input(body: ProjectInput) -> dict[str, str | None]:
    return {
        "name": body.name.strip(),
        "description": body.description.strip()
        if body.description and body.description.strip()
        else None,
    }


def project_request_hash(body: ProjectInput) -> bytes:
    encoded = json.dumps(
        normalized_project_input(body), ensure_ascii=False, separators=(",", ":"), sort_keys=True
    ).encode()
    return sha256(encoded).digest()


def project_json(project: Project) -> dict[str, object]:
    return {
        "id": project.id,
        "tenantId": project.tenant_id,
        "name": project.name,
        "description": project.description,
        "version": project.version,
        "createdAt": project.created_at.replace(tzinfo=UTC).isoformat(),
        "updatedAt": project.updated_at.replace(tzinfo=UTC).isoformat(),
    }


@router.get("")
async def list_projects(
    cursor_time: datetime | None = Query(default=None, alias="cursorTime"),
    cursor_id: str | None = Query(default=None, alias="cursorId"),
    limit: int = Query(default=50, ge=1, le=100),
    actor: Principal = Depends(current_principal),
    session: AsyncSession = Depends(database_session),
) -> dict[str, object]:
    statement = select(Project).where(
        Project.tenant_id == actor.tenant_id, Project.deleted_at.is_(None)
    )
    if cursor_time and cursor_id:
        statement = statement.where(
            or_(
                Project.updated_at < cursor_time,
                and_(Project.updated_at == cursor_time, Project.id < cursor_id),
            )
        )
    rows = list(
        (
            await session.scalars(
                statement.order_by(Project.updated_at.desc(), Project.id.desc()).limit(limit + 1)
            )
        ).all()
    )
    has_next = len(rows) > limit
    items = rows[:limit]
    next_cursor = (
        None
        if not has_next
        else {"updatedAt": items[-1].updated_at.isoformat(), "id": items[-1].id}
    )
    return {"items": [project_json(item) for item in items], "nextCursor": next_cursor}


@router.get("/{project_id}")
async def get_project(
    project_id: UUID,
    actor: Principal = Depends(current_principal),
    session: AsyncSession = Depends(database_session),
) -> dict[str, object]:
    project = await session.scalar(
        select(Project).where(
            Project.id == str(project_id),
            Project.tenant_id == actor.tenant_id,
            Project.deleted_at.is_(None),
        )
    )
    if project is None:
        raise ApiProblem(404, "resource_not_found", "Project is not visible")
    return project_json(project)


@router.post("", status_code=201)
async def create_project(
    body: ProjectInput,
    idempotency_key: str = Header(min_length=16, max_length=128),
    actor: Principal = Depends(current_principal),
    session: AsyncSession = Depends(database_session),
) -> dict[str, object]:
    request_hash = project_request_hash(body)
    normalized = normalized_project_input(body)
    now = datetime.now(UTC).replace(tzinfo=None)
    try:
        async with session.begin():
            existing = await session.scalar(
                select(IdempotencyKey)
                .where(
                    IdempotencyKey.tenant_id == actor.tenant_id,
                    IdempotencyKey.scope == IDEMPOTENCY_SCOPE,
                    IdempotencyKey.idem_key == idempotency_key,
                )
                .with_for_update()
            )
            if existing:
                if existing.request_hash != request_hash:
                    raise ApiProblem(
                        409, "idempotency_key_reused", "Idempotency key has another payload"
                    )
                if existing.status == "completed" and existing.response_json:
                    return existing.response_json
                raise ApiProblem(
                    409, "idempotency_request_in_progress", "Original request is still running"
                )

            idempotency = IdempotencyKey(
                tenant_id=actor.tenant_id,
                scope=IDEMPOTENCY_SCOPE,
                idem_key=idempotency_key,
                request_hash=request_hash,
                status="processing",
                response_status=None,
                response_json=None,
                expires_at=now + timedelta(hours=24),
                created_at=now,
            )
            project = Project(
                id=str(uuid4()),
                tenant_id=actor.tenant_id,
                owner_id=actor.user_id,
                name=normalized["name"],
                description=normalized["description"],
                version=1,
                created_at=now,
                updated_at=now,
            )
            session.add_all([idempotency, project])
            await session.flush()
            response = project_json(project)
            idempotency.status = "completed"
            idempotency.response_status = 201
            idempotency.response_json = response
        return response
    except IntegrityError as error:
        await session.rollback()
        existing = await session.scalar(
            select(IdempotencyKey).where(
                IdempotencyKey.tenant_id == actor.tenant_id,
                IdempotencyKey.scope == IDEMPOTENCY_SCOPE,
                IdempotencyKey.idem_key == idempotency_key,
            )
        )
        if existing:
            if existing.request_hash != request_hash:
                raise ApiProblem(
                    409, "idempotency_key_reused", "Idempotency key has another payload"
                ) from error
            if existing.status == "completed" and existing.response_json:
                return cast(dict[str, object], existing.response_json)
            raise ApiProblem(
                409, "idempotency_request_in_progress", "Original request is still running"
            ) from error
        raise ApiProblem(409, "project_name_exists", "Project already exists") from error


@router.patch("/{project_id}")
async def update_project(
    project_id: UUID,
    body: ProjectUpdate,
    actor: Principal = Depends(current_principal),
    session: AsyncSession = Depends(database_session),
) -> dict[str, object]:
    statement = (
        update(Project)
        .where(
            Project.id == str(project_id),
            Project.tenant_id == actor.tenant_id,
            Project.version == body.version,
            Project.deleted_at.is_(None),
        )
        .values(
            name=body.name,
            description=body.description,
            version=Project.version + 1,
            updated_at=datetime.now(UTC).replace(tzinfo=None),
        )
    )
    result = await session.execute(statement)
    if result.rowcount != 1:
        exists = await session.scalar(
            select(Project.id).where(
                Project.id == str(project_id),
                Project.tenant_id == actor.tenant_id,
                Project.deleted_at.is_(None),
            )
        )
        raise ApiProblem(
            409 if exists else 404,
            "version_conflict" if exists else "not_found",
            "Project changed or is not visible",
        )
    await session.commit()
    project = await session.scalar(
        select(Project).where(Project.id == str(project_id), Project.tenant_id == actor.tenant_id)
    )
    assert project is not None
    return project_json(project)


@router.delete("/{project_id}", status_code=204)
async def delete_project(
    project_id: UUID,
    actor: Principal = Depends(current_principal),
    session: AsyncSession = Depends(database_session),
) -> Response:
    result = await session.execute(
        update(Project)
        .where(
            Project.id == str(project_id),
            Project.tenant_id == actor.tenant_id,
            Project.deleted_at.is_(None),
        )
        .values(deleted_at=datetime.now(UTC).replace(tzinfo=None))
    )
    if result.rowcount != 1:
        raise ApiProblem(404, "not_found", "Project is not visible")
    await session.commit()
    return Response(status_code=204)

File: examples/backend/go/internal/api/projects.go
package api

import (
	"bytes"
	"crypto/sha256"
	"encoding/json"
	"errors"
	"net/http"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
	"gorm.io/gorm"
	"gorm.io/gorm/clause"
)

const projectCreateScope = "projects.create"

var (
	errIdempotencyReused     = errors.New("idempotency key has another payload")
	errIdempotencyInProgress = errors.New("original request is still running")
)

type Project struct {
	ID          string     `gorm:"column:id;primaryKey;type:char(36)" json:"id"`
	TenantID    string     `gorm:"column:tenant_id;type:char(36);index" json:"tenantId"`
	OwnerID     string     `gorm:"column:owner_id;type:char(36)" json:"-"`
	Name        string     `gorm:"column:name;size:120" json:"name"`
	Description *string    `gorm:"column:description;type:text" json:"description"`
	Status      string     `gorm:"column:status" json:"status"`
	Version     uint64     `gorm:"column:version" json:"version"`
	DeletedAt   *time.Time `gorm:"column:deleted_at" json:"-"`
	CreatedAt   time.Time  `gorm:"column:created_at" json:"createdAt"`
	UpdatedAt   time.Time  `gorm:"column:updated_at" json:"updatedAt"`
}

func (Project) TableName() string { return "projects" }

type projectInput struct {
	Name        *string `json:"name" binding:"omitempty,min=2,max=120"`
	Description *string `json:"description" binding:"omitempty,max=4000"`
	Version     uint64  `json:"version"`
}

type IdempotencyKey struct {
	TenantID       string          `gorm:"column:tenant_id;primaryKey"`
	Scope          string          `gorm:"column:scope;primaryKey"`
	Key            string          `gorm:"column:idem_key;primaryKey"`
	RequestHash    []byte          `gorm:"column:request_hash;type:binary(32)"`
	Status         string          `gorm:"column:status"`
	ResponseStatus *uint16         `gorm:"column:response_status"`
	ResponseJSON   json.RawMessage `gorm:"column:response_json;type:json"`
	ExpiresAt      time.Time       `gorm:"column:expires_at"`
	CreatedAt      time.Time       `gorm:"column:created_at"`
}

func (IdempotencyKey) TableName() string { return "idempotency_keys" }

type normalizedProjectInput struct {
	Name        string  `json:"name"`
	Description *string `json:"description"`
}

func normalizeProjectInput(input projectInput) normalizedProjectInput {
	name := strings.TrimSpace(*input.Name)
	var description *string
	if input.Description != nil {
		trimmed := strings.TrimSpace(*input.Description)
		if trimmed != "" {
			description = &trimmed
		}
	}
	return normalizedProjectInput{Name: name, Description: description}
}

func projectRequestHash(input normalizedProjectInput) ([32]byte, error) {
	encoded, err := json.Marshal(input)
	if err != nil {
		return [32]byte{}, err
	}
	return sha256.Sum256(encoded), nil
}

func idempotentProject(record IdempotencyKey, requestHash [32]byte) (*Project, error) {
	if !bytes.Equal(record.RequestHash, requestHash[:]) {
		return nil, errIdempotencyReused
	}
	if record.Status != "completed" || len(record.ResponseJSON) == 0 {
		return nil, errIdempotencyInProgress
	}
	var project Project
	if err := json.Unmarshal(record.ResponseJSON, &project); err != nil {
		return nil, err
	}
	return &project, nil
}

func (s *Server) listProjects(c *gin.Context) {
	actor := principalFrom(c)
	var projects []Project
	if err := s.db.WithContext(c.Request.Context()).
		Where("tenant_id = ? AND deleted_at IS NULL", actor.TenantID).
		Order("updated_at DESC, id DESC").Limit(50).Find(&projects).Error; err != nil {
		writeProblem(c, http.StatusServiceUnavailable, "database_unavailable", "Project query failed")
		return
	}
	c.JSON(http.StatusOK, gin.H{"items": projects, "nextCursor": nil})
}

func (s *Server) getProject(c *gin.Context) {
	actor := principalFrom(c)
	var project Project
	if err := s.db.WithContext(c.Request.Context()).
		Where("id = ? AND tenant_id = ? AND deleted_at IS NULL", c.Param("id"), actor.TenantID).
		First(&project).Error; err != nil {
		writeProblem(c, http.StatusNotFound, "resource_not_found", "Project is not visible")
		return
	}
	c.JSON(http.StatusOK, project)
}

func (s *Server) createProject(c *gin.Context) {
	actor := principalFrom(c)
	key := c.GetHeader("Idempotency-Key")
	if len(key) < 16 || len(key) > 128 {
		writeProblem(c, http.StatusUnprocessableEntity, "invalid_field", "Idempotency-Key is required")
		return
	}
	var input projectInput
	if err := c.ShouldBindJSON(&input); err != nil || input.Name == nil {
		writeProblem(c, http.StatusUnprocessableEntity, "invalid_field", "Project input is invalid")
		return
	}
	normalized := normalizeProjectInput(input)
	requestHash, err := projectRequestHash(normalized)
	if err != nil {
		writeProblem(c, http.StatusInternalServerError, "internal_error", "Unable to hash request")
		return
	}
	now := time.Now().UTC()
	project := Project{
		ID: uuid.NewString(), TenantID: actor.TenantID, OwnerID: actor.UserID,
		Name: normalized.Name, Description: normalized.Description, Status: "draft", Version: 1,
		CreatedAt: now, UpdatedAt: now,
	}
	err = s.db.WithContext(c.Request.Context()).Transaction(func(tx *gorm.DB) error {
		var existing IdempotencyKey
		lookup := tx.Clauses(clause.Locking{Strength: "UPDATE"}).Where(
			"tenant_id = ? AND scope = ? AND idem_key = ?", actor.TenantID, projectCreateScope, key,
		).First(&existing)
		if lookup.Error == nil {
			matched, matchError := idempotentProject(existing, requestHash)
			if matchError != nil {
				return matchError
			}
			project = *matched
			return nil
		}
		if !errors.Is(lookup.Error, gorm.ErrRecordNotFound) {
			return lookup.Error
		}

		idempotency := IdempotencyKey{
			TenantID: actor.TenantID, Scope: projectCreateScope, Key: key,
			RequestHash: requestHash[:], Status: "processing", ExpiresAt: now.Add(24 * time.Hour), CreatedAt: now,
		}
		if createError := tx.Create(&idempotency).Error; createError != nil {
			return createError
		}
		if createError := tx.Create(&project).Error; createError != nil {
			return createError
		}
		responseJSON, marshalError := json.Marshal(project)
		if marshalError != nil {
			return marshalError
		}
		status := uint16(http.StatusCreated)
		return tx.Model(&IdempotencyKey{}).Where(
			"tenant_id = ? AND scope = ? AND idem_key = ?", actor.TenantID, projectCreateScope, key,
		).Updates(map[string]any{
			"status": "completed", "response_status": status, "response_json": responseJSON,
		}).Error
	})
	if err != nil {
		if errors.Is(err, errIdempotencyReused) {
			writeProblem(c, http.StatusConflict, "idempotency_key_reused", err.Error())
			return
		}
		if errors.Is(err, errIdempotencyInProgress) {
			writeProblem(c, http.StatusConflict, "idempotency_request_in_progress", err.Error())
			return
		}
		var existing IdempotencyKey
		lookup := s.db.WithContext(c.Request.Context()).Where(
			"tenant_id = ? AND scope = ? AND idem_key = ?", actor.TenantID, projectCreateScope, key,
		).First(&existing)
		if lookup.Error == nil {
			matched, matchError := idempotentProject(existing, requestHash)
			switch {
			case matchError == nil:
				c.JSON(http.StatusCreated, matched)
			case errors.Is(matchError, errIdempotencyReused):
				writeProblem(c, http.StatusConflict, "idempotency_key_reused", matchError.Error())
			default:
				writeProblem(c, http.StatusConflict, "idempotency_request_in_progress", matchError.Error())
			}
			return
		}
		if !errors.Is(lookup.Error, gorm.ErrRecordNotFound) {
			writeProblem(c, http.StatusServiceUnavailable, "database_unavailable", "Project creation failed")
			return
		}
		writeProblem(c, http.StatusConflict, "project_name_exists", "Project already exists")
		return
	}
	c.JSON(http.StatusCreated, project)
}

func (s *Server) updateProject(c *gin.Context) {
	actor := principalFrom(c)
	var input projectInput
	if err := c.ShouldBindJSON(&input); err != nil || input.Version < 1 {
		writeProblem(c, http.StatusUnprocessableEntity, "invalid_field", "Project input is invalid")
		return
	}
	changes := map[string]any{"version": input.Version + 1, "updated_at": time.Now().UTC()}
	if input.Name != nil {
		changes["name"] = *input.Name
	}
	if input.Description != nil {
		changes["description"] = *input.Description
	}
	result := s.db.WithContext(c.Request.Context()).Model(&Project{}).
		Where("id = ? AND tenant_id = ? AND version = ? AND deleted_at IS NULL", c.Param("id"), actor.TenantID, input.Version).
		Updates(changes)
	if result.Error != nil {
		writeProblem(c, http.StatusServiceUnavailable, "database_unavailable", "Project update failed")
		return
	}
	if result.RowsAffected != 1 {
		var count int64
		s.db.WithContext(c.Request.Context()).Model(&Project{}).
			Where("id = ? AND tenant_id = ? AND deleted_at IS NULL", c.Param("id"), actor.TenantID).Count(&count)
		if count == 0 {
			writeProblem(c, http.StatusNotFound, "resource_not_found", "Project is not visible")
		} else {
			writeProblem(c, http.StatusConflict, "version_conflict", "Project version changed")
		}
		return
	}
	s.getProject(c)
}

func (s *Server) deleteProject(c *gin.Context) {
	actor := principalFrom(c)
	result := s.db.WithContext(c.Request.Context()).Model(&Project{}).
		Where("id = ? AND tenant_id = ? AND deleted_at IS NULL", c.Param("id"), actor.TenantID).
		Updates(map[string]any{"deleted_at": time.Now().UTC(), "version": gormExpr("version + 1")})
	if result.Error != nil {
		writeProblem(c, http.StatusServiceUnavailable, "database_unavailable", "Project delete failed")
		return
	}
	if result.RowsAffected != 1 {
		writeProblem(c, http.StatusNotFound, "resource_not_found", "Project is not visible")
		return
	}
	c.Status(http.StatusNoContent)
}
